pkgs/development/tools/rust/cargo-llvm-cov/default.nix

```diff
@@ -0,0 +1,63 @@
+{ lib
+, stdenv
+, fetchurl
+, fetchFromGitHub
+, rustPlatform
+}:
+
+# The dance with `fetchurl`, `fetchFromGitHub`, and `postPatch` is because
+# `fetchCrate` wouldn't pull in the fixtures needed to run the test suite.
+
+let
+  pname = "cargo-llvm-cov";
+  version = "0.5.0";
+
+  owner = "taiki-e";
+  repo = pname;
+
+  llvm = rustPlatform.rust.rustc.llvmPackages.llvm;
+
+  cargoLock = fetchurl {
+    name = "Cargo.lock";
+    url = "https://crates.io/api/v1/crates/${pname}/${version}/download";
+    sha256 = "sha256-ts/uL8VndeODDpmpTJZD049I7nOIiSkZJRONSFOD9Mg=";
+    downloadToTemp = true;
+    postFetch = ''
+      tar xzf $downloadedFile ${pname}-${version}/Cargo.lock
+      mv */Cargo.lock $out
+    '';
+  };
+in
+
+rustPlatform.buildRustPackage {
+  inherit pname version;
+
+  src = fetchFromGitHub {
+    inherit owner repo;
+    rev = "v${version}";
+    sha256 = "sha256-2O0MyL4SF/2AUpgWYUDWQ5dDpa84pwmnKGtAaWi5bwQ=";
+  };
+
+  cargoLock.lockFile = "${cargoLock}";
```

Comment:
This is an instance of [IFD](https://nixos.wiki/wiki/Import_From_Derivation), which is a no-go for Nixpkgs. Add the copy to `cargoUpdateHook`, and set `cargoVendorSha256` instead.

Reply:
Honestly I'm kinda lost here. I understand how this is IFD and why that's not allowed (but didn't realize till you pointed it out), but searching for those things in my local copy Nixpkgs doesn't give me anything useful to go off of. I've tried a bunch of random things but can't get past this check: https://github.com/NixOS/nixpkgs/blob/f157bef02b0902079561a02a2559e89ccb13236b/pkgs/build-support/rust/fetch-cargo-tarball/default.nix#L44-L46

Reply:
Right, because the Cargo update hook runs after that check, but before anything else, for whatever reason.

I personally don't see any downsides to moving it above that check -- do you mind making a commit to do that on this branch? Name it something like `fetchCargoTarball: run cargoUpdateHook before checking for Cargo.lock` or something, not sure. I'll ping someone re: that change after you push it just in case it's undesirable for whatever reason.

Reply:
Hmm, that does fix it. It's weird to me that `postPatch = "cp ${cargoLock} Cargo.lock";` isn't enough to pass the check on its own though, and it looks like that's required even after this change too.

Reply:
After this change, `nix-env -iA cargo-llvm-cov -f .` is spending a lot of time updating the crates.io index and downloading and vendoring crates, which the IFD version wasn't doing. I don't know if that means this change broke memoization completely or if it just busted an existing cache, but I thought it'd be worth mentioning.

Reply:
> Hmm, that does fix it. It's weird to me that `postPatch = "cp ${cargoLock} Cargo.lock";` isn't enough to pass the check on its own though, and it looks like that's required even after this change too.

This is because the Cargo tarball derivation doesn't inherit the `postPatch` attribute, at least when using `buildRustPackage`.

> After this change, `nix-env -iA cargo-llvm-cov -f .` is spending a lot of time updating the crates.io index and downloading and vendoring crates, which the IFD version wasn't doing. I don't know if that means this change broke memoization completely or if it just busted an existing cache, but I thought it'd be worth mentioning.

This is expected -- the memoization of the crate files requires IFD. The `cargoLock.lockFile` mechanism and the vendoring mechanism are separate ways of retrieving the dependencies, one using IFD (and allowing for memoization) and one not.

Reply:
> This is because the Cargo tarball derivation doesn't inherit the `postPatch` attribute, at least when using `buildRustPackage`.

Yeah, I understand why it's needed, I was just noting that the UX was a little weird.

> This is expected -- the memoization of the crate files requires IFD. The `cargoLock.lockFile` mechanism and the vendoring mechanism are separate ways of retrieving the dependencies, one using IFD (and allowing for memoization) and one not.

Okay cool, just wanted to make sure I didn't screw anything up :sweat_smile:

Reply:
> My understanding is that it has to be this way because of how vendoring works.

I honestly don't see what you mean -- as long as the patches result in the same lockfile being vendored as being added in the source, it shouldn't matter. Not sure why it was done like it is (e.g. separate `cargoPatches` array), sadly.

Reply:
if i understood correctly, we can also use postFetch here inside the fetchFromGitHub

Reply:
Yes, we can also do that.

However, I think we also shouldn't have this issue in the first place; ideally, the dependencies derivation would inherit the `postPatch` phase from the parent drv when using `buildRustPackage` -- or supply a separate option for it, like `cargoPatches`. But we have `cargoUpdateHook` so honestly I don't know. Maybe we should refactor how this works a bit? *shrug*

Reply:
So I've just swapped over to @figsoda's `postFetch` suggestion, and this also obviates the need for the changes to `fetchCargoTarball`. I've left that commit in though because we probably want to do it anyway. Let me know if we want to drop it since it's not required anymore though.

> However, I think we also shouldn't have this issue in the first place; ideally, the dependencies derivation would inherit the `postPatch` phase from the parent drv when using `buildRustPackage` -- or supply a separate option for it, like `cargoPatches`. But we have `cargoUpdateHook` so honestly I don't know. Maybe we should refactor how this works a bit? _shrug_

I am in agreement with all of this, lol